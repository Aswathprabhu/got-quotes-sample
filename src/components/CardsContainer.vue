<template>
  <div class="card-container">
    <Card
      v-for="index in Number(count)"
      :key="index"
      :data="result[index - 1]"
    />
  </div>
</template>

<script>
import 'isomorphic-fetch';
import Card from './Card.vue';
export default {
  name: 'CardsContainer',
  components: {
    Card,
  },
  data() {
    return {
      result: [],
    };
  },
  created() {
    this.fetchResults(this.count);
  },
  computed: {
    count() {
      return this.$route.params.id || 1;
    },
  },
  watch: {
    count(val) {
      this.fetchResults(val);
    },
  },
  methods: {
    fetchResults(count) {
      return fetch(
        `https://got-quotes-api.vercel.app/api/random/${count}`
      )
        .then((res) => res.json())
        .then((res) => (this.result = Array.isArray(res) ? res : [res]))
        .catch(console.error);
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.card-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}
</style>
