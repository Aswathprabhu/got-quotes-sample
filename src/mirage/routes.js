export default function getRoutes() {
  this.get('/random/:id', (schema, request) => {
    let count = request.params.id || 1;
    return schema.quotes.find(
      Array.from(Array(Number(count)), (_, idx) => idx + 1)
    );
  });
}
