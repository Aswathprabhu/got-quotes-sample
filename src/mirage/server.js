import { createServer, Factory, Model, RestSerializer } from 'miragejs';
import getRoutes from './routes';

const isTestEnv = (env) => env === 'test';

export function makeServer({ environment = 'test' }) {
  return createServer({
    environment,
    models: {
      quote: Model,
    },
    factories: {
      quote: Factory.extend({
        sentence: 'The man who passes the sentence should swing the sword.',
        character: {
          name: 'Eddard "Ned" Stark',
          slug: 'ned',
          house: {
            name: 'House Stark of Winterfell',
            slug: 'stark',
          },
        },
      }),
    },
    seeds(server) {
      server.createList('quote', 20);
    },
    routes() {
      this.passthrough('https://game-of-thrones-quotes.herokuapp.com/**');
      this.urlPrefix = 'https://game-of-thrones-quotes.herokuapp.com';
      this.namespace = '/v1';
      isTestEnv(environment) && getRoutes.bind(this)();
    },
    serializers: {
      application: RestSerializer.extend({
        // https://miragejs.com/api/classes/serializer/#embed
        embed: true,
        // https://miragejs.com/api/classes/serializer/#root
        root: false,
      }),
    },
  });
}
