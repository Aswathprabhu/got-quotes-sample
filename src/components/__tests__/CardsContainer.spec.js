import CardsContainer from '../CardsContainer.vue';
import { render, waitForElementToBeRemoved } from '@testing-library/vue';
import { makeServer } from '../../mirage/server';

describe('CardsContainer - Initial Render', () => {
  let server;

  beforeEach(() => {
    server = makeServer({ environment: 'test' });
  });

  afterEach(() => {
    server.shutdown();
  });

  it('Initial Render', async () => {
    server.createList('quote', 2);
    const { getAllByTestId } = render(CardsContainer, {
      mocks: {
        $route: {
          params: { id: '2' },
        },
      },
    });
    let loaders = getAllByTestId('loader');
    expect(loaders.length).toBe(2);
  });

  it('After Data', async () => {
    server.createList('quote', 2);
    const { getAllByTestId } = render(CardsContainer, {
      mocks: {
        $route: {
          params: { id: '2' },
        },
      },
    });
    await waitForElementToBeRemoved(() => getAllByTestId('loader')[0]);
    const dataCards = getAllByTestId('data-card');
    expect(dataCards.length).toBe(2);
  });

  it('Check if UI doesnt break for invalid data', async () => {
    server.createList('quote', 1, { character: {}, sentence: '' });
    const { getAllByTestId, getAllByText, container } = render(CardsContainer, {
      mocks: {
        $route: {
          params: { id: '1' },
        },
      },
    });
    await waitForElementToBeRemoved(() => getAllByTestId('loader')[0]);
    expect(container).toMatchSnapshot();
    const dataCards = getAllByTestId('data-card');
    expect(dataCards.length).toBe(1);
    expect(getAllByText('N/A').length).toBe(3);
  });
});
