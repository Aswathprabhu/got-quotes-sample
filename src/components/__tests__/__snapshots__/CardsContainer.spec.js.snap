// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer - Initial Render Check if UI doesnt break for invalid data 1`] = `
<div>
  <div
    class="card-container"
  >
    <div
      class="card"
    >
      <div
        class=""
        data-testid="data-card"
      >
        <div>
          <h3>
            <u>
              Name
            </u>
          </h3>
           
          <h2>
            <i>
              N/A
            </i>
          </h2>
        </div>
         
        <div>
          <h3>
            <u>
              House
            </u>
          </h3>
           
          <h2>
            <i>
              N/A
            </i>
          </h2>
        </div>
         
        <div>
          <h3>
            <u>
              Quote
            </u>
          </h3>
           
          <h2>
            <i>
              N/A
            </i>
          </h2>
        </div>
      </div>
    </div>
  </div>
</div>
`;
